cadc-storage-adapter/src/main/java/org/opencadc/inventory/storage/StorageClient.java

```diff
@@ -0,0 +1,195 @@
+import ca.nrc.cadc.net.ResourceNotFoundException;
+import ca.nrc.cadc.net.TransientException;
+
+/**
+ * Provides access to storage.  
+ *
+ * @author majorb
+ *
+ */
+public class StorageClient {
+    
+    private static Logger log = Logger.getLogger(StorageClient.class);
+    
+    private StorageAdapter adapter;
+
+    public StorageClient() {
+        adapter = getStorageAdapter();
+    }
+    
+    public Artifact get(URI storageID, OutputStream out) throws ResourceNotFoundException, TransientException {
+        InputStreamWrapper handler = new InputStreamWrapper() {
+            public void read(InputStream in) throws IOException {
+                ioLoop(out, in);
+            }
+        };
+        return adapter.get(storageID, handler);
+    }
+
+    public Artifact meta(URI storageID) throws ResourceNotFoundException, TransientException {
+        return adapter.meta(storageID);
+    }
+
+    public StorageLocation put(Artifact artifact, InputStream in) throws StreamCorruptedException, TransientException {
+        OutputStreamWrapper wrapper = new OutputStreamWrapper() {
+            public void write(OutputStream out) throws IOException {
+                ioLoop(out, in);
+            }
+        };
+        return adapter.put(artifact, wrapper);
+    }
+
+    public StorageLocation replace(Artifact artifact, InputStream in, StorageLocation replaceID) throws ResourceNotFoundException, StreamCorruptedException, TransientException {
+        OutputStreamWrapper wrapper = new OutputStreamWrapper() {
+            public void write(OutputStream out) throws IOException {
+                ioLoop(out, in);
+            }
+        };
+        return adapter.replace(artifact, wrapper, replaceID);
+    }
+
+    public void delete(URI storageID) throws ResourceNotFoundException, TransientException {
+        adapter.delete(storageID);
+    }
+
+    public Iterator<StorageLocation> iterator() throws TransientException {
+        return adapter.iterator();
+    }
+    
+    private void ioLoop(OutputStream out, InputStream in) {
+        // TODO: Write 2 threaded io loop--one thread reading
+        // and one thread writing.
+    }
+    
+    /**
+     * Load the configured storage adapter or the default one if none
+     * are found.
+     * 
+     * @return The storage adapter to be used by this client.
+     */
+    private StorageAdapter getStorageAdapter() {
+        Class c = null;
+        
+        // Try to load the adapter based on a classname in a system property
+        String cname = System.getProperty(StorageAdapter.class.getName());
+        if (cname != null) {
+            try {
+                c = Class.forName(cname);
+            } catch (Throwable t) {
+                String msg = "Failed to find storage adapter " + cname;
+                log.error(msg, t);
+                throw new IllegalStateException(msg, t);
+            }
+        }
+        
+        // Try to load the adapter based on the default name
```

Comment:
This default class name approach won't be usable as it requires injecting new code into the war and we want to have a single reusable docker image.